### PR TITLE
fastmath: fast float32 exp for fused softmax

### PR DESCRIPTION
**Pull request by suyashgoel**
## Summary

Speeds up the fused softmax float32 kernel by 35–42% (per tensor shape) by
replacing `math.Exp` with a fast float32 exp approximation, with no change to
memory use and no compliance regressions, with approximation error below the
project’s tolerance.

## Motivation

The go-backend has no JIT, so fused kernels are hand-written. Profiling the
fused softmax kernel showed `math.Exp` dominated its runtime — softmax's
middle pass calls it once per element, and `math.Exp` operates in float64 (with
the float32 kernel paying a convert-in/convert-out cost on top).

Softmax doesn't need full `math.Exp` precision: it evaluates `exp(x - max)`, so
every argument is `≤ 0`, and the result is immediately normalized. A float32
approximation with `< 1e-7` relative error is well within tolerance.

## What changed

- New `internal/fastmath` package with `Exp32`, a Cephes-style `expf`
  approximation (range reduction `e^x = 2^n · e^g` + a degree-7 Horner
  polynomial, with the `2^n` scale built directly from the float32 exponent
  bits). Maximum relative error is below `1e-7` versus `math.Exp` over
  `[-87, 0]`, covering the normal-output range relevant to softmax.
- `fusedSoftmax` takes an `expFn` parameter; the float32 path uses
  `fastmath.Exp32`, the float64 path is unchanged (`math.Exp`).

## Benchmarks

`BenchmarkSoftmax/Fused`, Apple M1 (`darwin/arm64`), interleaved
main-vs-branch. The table reports the median of 8 runs; statistical
significance was evaluated with `benchstat`:

| shape | before | after | Δ |
|---|---|---|---|
| 8×64 | 13.43µs | 8.68µs | −35.4% |
| 32×128 | 78.53µs | 46.96µs | −40.2% |
| 64×512 | 589.9µs | 340.1µs | −42.3% |
| 8×16×64 | 156.5µs | 92.16µs | −41.1% |
| 4×8×32×128 | 2.414ms | 1.398ms | −42.1% |

All Δ at `p < 0.001` (`n=8`). `B/op` and `allocs/op` unchanged. Measured
end-to-end through the backend + executor, so it reflects caller-visible time,
not an isolated microbenchmark.

## Correctness

- Fused softmax compliance suite (`TestCompliance/FusedOps/FusedSoftmax`:
  1D, 2D, NegativeAxis, Stability) passes unchanged.
- New `Exp32` unit tests: exactness at 0, clamp behavior, finite/non-negative
  over `[-100, 100]` (guards an underflow edge case), and relative error
  `< 1e-7` vs `math.Exp`.

## Attribution

The float32 exp is derived from Cephes `expf` (Stephen L. Moshier) — the
range-reduction constants and polynomial coefficients are the standard Cephes
values, reimplemented in Go and credited in the source.

### CODE REVIEW

File: internal/fastmath/exp.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"math"
+)
+
+// Float32 range limits for exp, matching Cephes expf.
+//
+// Outside [minLogF, maxLogF], a finite float32 exp either overflows the
+// normal range or underflows to a subnormal/zero. We clamp instead of
+// returning ±Inf / subnormals: overflow → maxNumF, underflow → 0.
+const (
```
Comment (janpfeifer) [inline internal/fastmath/exp.go]: All these constants are for the Cephes `exp32` implementation only. And I imagine with time we will see more functions in the `fastmath` package. What about scoping these constants inside the `Exp32` or `expPoly` functions, where they are used to avoid naming conflict ?

Either that or prefixing them with `cephesExp32` (or something like that).

File: internal/fastmath/exp.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	// Exact power-of-two scale: float32 exponent bits hold n + bias(127).
+	// Safe because the clamps keep n ∈ [-127, 127].
+	scale := math.Float32frombits(uint32(int32(n)+127) << 23)
+	return scale * expPoly(g)
+}
+
+// expPoly approximates e^x for |x| ≤ ln2/2 using Cephes's Horner polynomial:
+//
+//	e^x ≈ 1 + x + x²·P(x)
+func expPoly(x float32) float32 {
```
Comment (janpfeifer) [inline internal/fastmath/exp.go]: What about `exp32Poly` to match the name of `Exp32` ? 

In case we decide to do a faster `Exp64` later ?

**Review by janpfeifer — COMMENTED** [review]
Very clean implementation, thanks!

I left a couple of comments to prevent future naming conflicts (we should not reserve a global package symbol like `p2`, if we expect the package to grow, other functions may want to use it).

But your PR inspired me to investigate a bit with the AI, and it suggested using the Schraudolph Method for exponentiation, albeit less precise, which it says is even faster. Very optional, but just to say that maybe we should suffix the `Exp32` with something, like `Exp32Cephes`, in expectation for a later `Exp32Schraudolph` ?

The Schraudolph method is less precise, but usually good enough for ML. It could be enabled with an option like `GOMLX_BACKEND="go:fastmath=true"` or something like that. Anyway, something for a future PR.

Food for thought: in the future we want to use the new Go SIMD to implement these as a vector (with 1 order of magnitude speedups). Not for this PR, but just to think about.

File: internal/fastmath/exp.go (hunk 1 above)
Comment (suyashgoel) [reply internal/fastmath/exp.go]: Makes sense, thanks!

**Comment by suyashgoel** [comment]
Thanks for the feedback! Scoped the Cephes consts into Exp32 / exp32Poly and renamed the helper to exp32Poly. Thinking to defer Exp32 → Exp32Cephes rename until we actually add another approx.

**Review by janpfeifer — COMMENTED** [review]
Thanks, perfect!
